### PR TITLE
prep AssignmentInfo to handle deconstructed-assignments

### PR DESCRIPTION
**Pull request by omsmith**
Ref: https://github.com/Brightspace/D2L.CodeStyle/issues/738

### CODE REVIEW

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 1]
```diff
@@ -352,9 +382,7 @@ ExpressionSyntax initializer
 			AssignmentExpressionSyntax assignmentSyntax,
 			ISymbol memberSymbol
 		) {
-			var semanticModel = m_compilation.GetSemanticModel(
-				memberSymbol.DeclaringSyntaxReferences.Single().GetSyntax().SyntaxTree
-			);
+			var semanticModel = m_compilation.GetSemanticModel( assignmentSyntax.SyntaxTree );
```
Comment (omsmith) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:385]: Err, extra hunk, but correcter change

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 2]
```diff
@@ -301,16 +301,38 @@ out diagnostic
...
 
 			public bool IsInitializer { get; }
 			public ExpressionSyntax Expression { get; }
+			public ITypeSymbol AssignedType { get; }
+
+			public static AssignmentInfo Create(
+				SemanticModel model,
+				bool isInitializer,
+				ExpressionSyntax expression
+			) {
```
Comment (omsmith) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:322]: Might not end up with this, ~important change is it now has `AssignedType` exposed

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 3]
```diff
@@ -432,7 +459,7 @@ private enum AssignmentQueryKind {
 					// Some methods are known to have return values that are
 					// immutable (such as Enumerable.Empty()).
 					// These should be considered immutable by the Analyzer.
-					semanticModel ??= m_compilation.GetSemanticModel( assignment.Expression.SyntaxTree );
+					SemanticModel semanticModel = m_compilation.GetSemanticModel( assignment.Expression.SyntaxTree );
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:462]: I wonder if the symbol info should be part of `AssignmentInfo`, as we are passing in the same semantic model to the `Create`

**Review by j3parker — APPROVED** [review]

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs (hunk 3 above)
Comment (omsmith) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:462]: Yeah - somehow not calling `GetSemanticModel` everywhere would be good - they cache information for the syntax tree internally. At the same time, don't want to hold onto them longer than necessary, but that's probably the same lifetime as the syntax node itself anyway (as opposed to a syntax reference).

**Pull request merged**